File: assets/Scripts/player/PlayerPawn.js
// FILE: Scripts/player/PlayerPawn.js
"use strict";

const U = require("./util.js");
const FrameContext = require("./FrameContext.js");
const CharacterConfig = require("./CharacterConfig.js");
const {PlayerEntityFactory} = require("./PlayerEntityFactory.js");
const InputRouter = require("./systems/InputRouter.js");

function req(v, msg) {
    if (v == null) throw new Error(msg);
    return v;
}

function apiFrom(ctx) {
    if (ctx && typeof ctx.api === "function") return ctx.api();
    if (ctx && ctx.engine && typeof ctx.engine.api === "function") return ctx.engine.api();
    if (ctx && typeof ctx.engineApi === "function") return ctx.engineApi();
    throw new Error("[player] ctx must provide api()");
}

function buildDomains(ctx) {
    const E = apiFrom(ctx);

    const ENGINE = req(globalThis.ENGINE, "[player] globalThis.ENGINE is required");
    const physics = req(ENGINE.physics, "[player] ENGINE.physics is required");

    const input = req(E.input && E.input(), "[player] engine.input() required");
    const camera = req(E.camera && E.camera(), "[player] engine.camera() required");
    const assets = req(E.assets && E.assets(), "[player] engine.assets() required");

    const entity = req(E.entity && E.entity(), "[player] engine.entity() required");
    const mesh = req(E.mesh && E.mesh(), "[player] engine.mesh() required");
    const surface = req(E.surface && E.surface(), "[player] engine.surface() required");

    const hud = req((typeof HUD !== "undefined" && HUD) ? HUD : null, "[player] HUD builtin required");
    const hudNative = (typeof E.hud === "function") ? E.hud() : null;

    const bus = (typeof E.bus === "function") ? E.bus() : null;

    return Object.freeze({
        ctx,
        engine: E,
        physics,
        input,
        camera,
        assets,
        entity,
        mesh,
        surface,
        bus,
        hud,
        hudNative
    });
}

function readUuidFromHandle(h) {
    if (!h) return "";
    if (typeof h.uuidString === "function") return String(h.uuidString() || "");
    if (typeof h.uuid === "function") return String(h.uuid() || "");
    if (typeof h.uuid === "string") return String(h.uuid || "");
    return "";
}

class PlayerPawn {
    constructor(ctx, cfg) {
        this.ctx = req(ctx, "[PlayerPawn] ctx required");
        this.cfg = cfg || Object.create(null);

        this.d = null;

        this.characterCfg = new CharacterConfig();
        this.frame = new FrameContext();
        this.inputRouter = null;

        this.handle = null; // EntityHandle (authoritative access)
        this.core = null;   // Optional UI mirror only

        this.alive = false;

        this._groundProbe = null;
    }

    get entity() {
        return this.handle;
    }

    get uuid() {
        return readUuidFromHandle(this.handle);
    }

    get bodyAccess() {
        const h = this.handle;
        if (!h) throw new Error("[PlayerPawn] bodyAccess on null handle");
        if (typeof h.bodyRef !== "function") {
            throw new Error("[PlayerPawn] EntityHandle.bodyRef() required (canonical physics access)");
        }
        return h.bodyRef();
    }

    get bodyId() {
        const h = this.handle;
        if (!h) return 0;
        if (typeof h.requireBodyId === "function") return h.requireBodyId("PlayerPawn.bodyId");
        if (typeof h.hasBody === "function" && h.hasBody() && typeof h.bodyRef === "function") {
            // best-effort: bodyRef exists but no id accessor -> treat as contract violation
            throw new Error("[PlayerPawn] EntityHandle.requireBodyId() required");
        }
        return 0;
    }

    get surfaceId() {
        const h = this.handle;
        if (!h) return 0;
        // If your EntityHandle exposes surfaceId directly, use it.
        if (typeof h.surfaceId === "number") return h.surfaceId | 0;

        // Otherwise derive from snapshot "binding" (recommended).
        if (typeof h.snapshot === "function") {
            const snap = h.snapshot();
            const byName = snap && (snap.componentsByName || snap.components) || null;
            const binding = byName ? byName.binding : null;
            return binding ? (binding.surfaceId | 0) : 0;
        }
        return 0;
    }

    get state() {
        // Optional: if you still need "state", store it in ECS component and read from snapshot.
        const h = this.handle;
        if (!h || typeof h.snapshot !== "function") return null;
        const snap = h.snapshot();
        const byName = snap && (snap.componentsByName || snap.components) || null;
        return byName ? (byName.state || null) : null;
    }

    init() {
        if (this.alive) return this;

        this.cfg = U.deepMerge({
            character: {radius: 0.35, height: 1.80, mass: 80.0, eyeHeight: 1.65},
            spawn: {pos: {x: 135, y: -10, z: -334}, radius: 0.35, height: 1.80, mass: 80.0},
            camera: {type: "third"},
            ui: {},
            events: {enabled: true},
            input: {},
            movement: {},
            shoot: {}
        }, this.cfg);

        this.d = buildDomains(this.ctx);

        this.inputRouter = new InputRouter(this.d.input, this.cfg.input);

        const factory = new PlayerEntityFactory(this);
        this.handle = factory.create(this.cfg.spawn);

        const uuid = this.uuid;
        if (!uuid) throw new Error("[PlayerPawn] player uuid missing (UUID-only)");

        // Optional UI mirror: keep if factory provides it, но НЕ требуем.
        this.core = this.handle && this.handle.core ? this.handle.core : null;
        if (this.core && typeof this.core === "object") {
            this.core.uuid = uuid;
        }

        // Canonical physics access validation
        if (typeof this.handle.bodyRef !== "function") {
            throw new Error("[PlayerPawn] EntityHandle.bodyRef() required");
        }
        if (typeof this.handle.hasBody === "function" && !this.handle.hasBody()) {
            throw new Error("[PlayerPawn] player entity has no physics body");
        }
        // requireBodyId is preferred (deterministic)
        if (typeof this.handle.requireBodyId === "function") {
            const id = this.handle.requireBodyId("PlayerPawn.init");
            if ((id | 0) <= 0) throw new Error("[PlayerPawn] invalid bodyId");
        }

        if (typeof this.frame.probeGroundCapsule !== "function") {
            throw new Error("[PlayerPawn] FrameContext.probeGroundCapsule required");
        }

        this.characterCfg.loadFrom(this.cfg, this.cfg.movement);

        // Ground probe uses canonical bodyAccess/bodyId
        this._groundProbe = () => {
            const probe = this.frame.probeGroundCapsule;
            const ba = this.bodyAccess;
            const bid = this.bodyId;

            return (probe.length >= 3)
                ? probe.call(this.frame, ba, this.characterCfg, bid)
                : probe.call(this.frame, ba, this.characterCfg);
        };

        this.alive = true;
        return this;
    }

    beginFrame(tpf) {
        if (!this.alive) throw new Error("[PlayerPawn] beginFrame on dead pawn");
        if (!Number.isFinite(tpf)) throw new Error("[PlayerPawn] tpf must be finite");

        const snap = this.d.input.consumeSnapshot();
        this.frame.begin(this, tpf, snap);

        this.inputRouter.read(this.frame);

        this.frame.bodyAccess = this.bodyAccess;
        this.frame.bodyId = this.bodyId;
    }

    syncPose() {
        if (!this.alive) throw new Error("[PlayerPawn] syncPose on dead pawn");

        // If FrameContext expects syncPhysics() on core, replace with local sampling:
        // - position/velocity/rotation/angularVelocity from bodyAccess
        // - grounded from ground probe
        const ba = this.bodyAccess;

        const p = ba.position();
        const v = ba.velocity();
        const r = ba.rotation();
        const av = ba.angularVelocity();

        const grounded = this._groundProbe ? !!this._groundProbe() : false;

        const pose = this.frame.pose;

        pose.x = +p.x;
        pose.y = +p.y;
        pose.z = +p.z;
        pose.vx = +v.x;
        pose.vy = +v.y;
        pose.vz = +v.z;
        pose.speed = Math.sqrt(pose.vx * pose.vx + pose.vz * pose.vz);
        pose.fallSpeed = (pose.vy < 0) ? -pose.vy : 0;

        pose.rx = +r.x;
        pose.ry = +r.y;
        pose.rz = +r.z;
        pose.rw = +r.w;
        pose.avx = +av.x;
        pose.avy = +av.y;
        pose.avz = +av.z;

        pose.grounded = grounded;
    }

    endFrame() {
        if (!this.alive) throw new Error("[PlayerPawn] endFrame on dead pawn");
        this.d.input.endFrame();
    }

    destroy() {
        const h = this.handle;
        this.handle = null;
        this.core = null;
        this.alive = false;

        if (h && typeof h.destroy === "function") {
            h.destroy();
        }
    }

    getModel() {
        // Если раньше это было core.model(), теперь либо:
        // 1) хранить модель в ECS/snapshot и читать, либо
        // 2) получать через surfaceId/SurfaceApi.
        if (this.core && typeof this.core.model === "function") return this.core.model();
        return null;
    }

    getBodyId() {
        return this.bodyId;
    }

    getSurfaceId() {
        return this.surfaceId;
    }

    getUuid() {
        return this.uuid;
    }
}

module.exports = {PlayerPawn};

File: assets/Scripts/player/PlayerPawn.lua
local M = {}
local luaRuntime = require("@builtin/lua_runtime")
local Error = luaRuntime.Error
local LuaConstruct = luaRuntime.LuaConstruct
local LuaClass = luaRuntime.LuaClass
local LuaNumberIsFinite = luaRuntime.LuaNumberIsFinite
local LuaNumber = luaRuntime.LuaNumber
local LuaDefineProperty = luaRuntime.LuaDefineProperty
U = require("./util.lua")
FrameContext = require("./FrameContext.lua")
CharacterConfig = require("./CharacterConfig.lua")
local lua_require_result_0 = require("./PlayerEntityFactory.lua")
PlayerEntityFactory = lua_require_result_0.PlayerEntityFactory
InputRouter = require("./systems/InputRouter.lua")
function req(self, v, msg)
    if v == nil then
        error(
            LuaConstruct(Error, msg),
            0
        )
    end
    return v
end
function apiFrom(self, ctx)
    if ctx and KTypeOf(ctx.api) == "function" then
        return ctx:api()
    end
    if ctx and ctx.engine and KTypeOf(ctx.engine.api) == "function" then
        return ctx.engine:api()
    end
    if ctx and KTypeOf(ctx.engineApi) == "function" then
        return ctx:engineApi()
    end
    error(
        LuaConstruct(Error, "[player] ctx must provide api()"),
        0
    )
end
function buildDomains(self, ctx)
    local E = apiFrom(_G, ctx)
    local ENGINE = req(_G, _G.ENGINE, "[player] _G.ENGINE is required")
    local physics = req(_G, ENGINE.physics, "[player] ENGINE.physics is required")
    local input = req(
        _G,
        E.input and E:input(),
        "[player] engine.input() required"
    )
    local camera = req(
        _G,
        E.camera and E:camera(),
        "[player] engine.camera() required"
    )
    local assets = req(
        _G,
        E.assets and E:assets(),
        "[player] engine.assets() required"
    )
    local entity = req(
        _G,
        E.entity and E:entity(),
        "[player] engine.entity() required"
    )
    local mesh = req(
        _G,
        E.mesh and E:mesh(),
        "[player] engine.mesh() required"
    )
    local surface = req(
        _G,
        E.surface and E:surface(),
        "[player] engine.surface() required"
    )
    local hud = req(_G, ENGINE.hud, "[player] ENGINE.hud module required")
    local lua_temp_4
    if KTypeOf(E.hud) == "function" then
        lua_temp_4 = E:hud()
    else
        lua_temp_4 = nil
    end
    local hudNative = lua_temp_4
    local lua_temp_5
    if KTypeOf(E.bus) == "function" then
        lua_temp_5 = E:bus()
    else
        lua_temp_5 = nil
    end
    local bus = lua_temp_5
    return KObject:freeze({
        ctx = ctx,
        engine = E,
        physics = physics,
        input = input,
        camera = camera,
        assets = assets,
        entity = entity,
        mesh = mesh,
        surface = surface,
        bus = bus,
        hud = hud,
        hudNative = hudNative
    })
end
function readUuidFromHandle(self, h)
    if not h then
        return ""
    end
    if KTypeOf(h.uuidString) == "function" then
        return tostring(h:uuidString() or "")
    end
    if KTypeOf(h.uuid) == "function" then
        return tostring(h:uuid() or "")
    end
    if KTypeOf(h.uuid) == "string" then
        return tostring(h.uuid or "")
    end
    return ""
end
PlayerPawn = LuaClass()
PlayerPawn.name = "PlayerPawn"
function PlayerPawn.prototype.lua_constructor(self, ctx, cfg)
    self.ctx = req(_G, ctx, "[PlayerPawn] ctx required")
    self.cfg = cfg or KObject:create(nil)
    self.d = nil
    self.characterCfg = LuaConstruct(CharacterConfig)
    self.frame = LuaConstruct(FrameContext)
    self.inputRouter = nil
    self.handle = nil
    self.core = nil
    self.alive = false
    ENGINE.input:grabMouse(true)
end
function PlayerPawn.prototype.init(self)
    if self.alive then
        return self
    end
    self.cfg = U:deepMerge({
        character = {radius = 0.35, height = 1.8, mass = 80, eyeHeight = 1.65},
        spawn = {pos = {x = 135, y = -10, z = -334}, radius = 0.35, height = 1.8, mass = 80},
        camera = {type = "third"},
        ui = {},
        events = {enabled = true},
        input = {},
        movement = {},
        shoot = {}
    }, self.cfg)
    self.d = buildDomains(_G, self.ctx)
    self.inputRouter = LuaConstruct(InputRouter, self.d.input, self.cfg.input)
    local factory = LuaConstruct(PlayerEntityFactory, self)
    self.handle = factory:create(self.cfg.spawn)
    local uuid = self.uuid
    if not uuid then
        error(
            LuaConstruct(Error, "[PlayerPawn] player uuid missing (UUID-only)"),
            0
        )
    end
    self.core = self.handle.core
    if not self.core then
        error(
            LuaConstruct(Error, "[PlayerPawn] ENGINE.entity.create() must return {core}"),
            0
        )
    end
    if not self.core.bodyAccess then
        error(
            LuaConstruct(Error, "[PlayerPawn] core.bodyAccess missing (engine must fill EntityCore)"),
            0
        )
    end
    if bit32.bor(self.core.bodyId, 0) <= 0 then
        error(
            LuaConstruct(Error, "[PlayerPawn] invalid core.bodyId"),
            0
        )
    end
    self.core.uuid = uuid
    if KTypeOf(self.frame.probeGroundCapsule) ~= "function" then
        error(
            LuaConstruct(Error, "[PlayerPawn] FrameContext.probeGroundCapsule required"),
            0
        )
    end
    self.characterCfg:loadFrom(self.cfg, self.cfg.movement)
    self.core:setGroundProbe(function(lua_, core)
        local probe = self.frame.probeGroundCapsule
        return KFunction:call(
            probe,
            self.frame,
            core.bodyAccess,
            self.characterCfg,
            bit32.bor(core.bodyId, 0)
        )
    end)
    self.alive = true
    return self
end
function PlayerPawn.prototype.beginFrame(self, tpf)
    if not self.alive then
        error(
            LuaConstruct(Error, "[PlayerPawn] beginFrame on dead pawn"),
            0
        )
    end
    if not LuaNumberIsFinite(tpf) then
        error(
            LuaConstruct(Error, "[PlayerPawn] tpf must be finite"),
            0
        )
    end
    local snap = self.d.input:consumeSnapshot()
    self.frame:begin(self, tpf, snap)
    self.inputRouter:read(self.frame)
    self.frame.bodyAccess = self.core.bodyAccess
    self.frame.bodyId = bit32.bor(self.core.bodyId, 0)
end
function PlayerPawn.prototype.syncPose(self)
    if not self.alive then
        error(
            LuaConstruct(Error, "[PlayerPawn] syncPose on dead pawn"),
            0
        )
    end
    local s = self.core:syncPhysics()
    local pose = self.frame.pose
    pose.x = s.x
    pose.y = s.y
    pose.z = s.z
    pose.vx = s.vx
    pose.vy = s.vy
    pose.vz = s.vz
    pose.speed = s.speed
    local lua_temp_6
    if s.vy < 0 then
        lua_temp_6 = LuaNumber(-s.vy)
    else
        lua_temp_6 = 0
    end
    pose.fallSpeed = lua_temp_6
    pose.rx = s.rx
    pose.ry = s.ry
    pose.rz = s.rz
    pose.rw = s.rw
    pose.avx = s.avx
    pose.avy = s.avy
    pose.avz = s.avz
    pose.grounded = s.grounded
end
function PlayerPawn.prototype.endFrame(self)
    if not self.alive then
        error(
            LuaConstruct(Error, "[PlayerPawn] endFrame on dead pawn"),
            0
        )
    end
    self.d.input:endFrame()
end
function PlayerPawn.prototype.destroy(self)
end
function PlayerPawn.prototype.getModel(self)
    return self.core:model()
end
function PlayerPawn.prototype.getBodyId(self)
    return bit32.bor(self.core.bodyId, 0)
end
function PlayerPawn.prototype.getSurfaceId(self)
    return bit32.bor(self.core.surfaceId, 0)
end
function PlayerPawn.prototype.getUuid(self)
    return self.uuid
end
LuaDefineProperty(
    PlayerPawn.prototype,
    "entity",
    {get = function(self)
        return self.handle
    end},
    true
)
LuaDefineProperty(
    PlayerPawn.prototype,
    "uuid",
    {get = function(self)
        return readUuidFromHandle(_G, self.handle)
    end},
    true
)
LuaDefineProperty(
    PlayerPawn.prototype,
    "bodyAccess",
    {get = function(self)
        return self.core.bodyAccess
    end},
    true
)
LuaDefineProperty(
    PlayerPawn.prototype,
    "bodyId",
    {get = function(self)
        return bit32.bor(self.core.bodyId, 0)
    end},
    true
)
LuaDefineProperty(
    PlayerPawn.prototype,
    "surfaceId",
    {get = function(self)
        return bit32.bor(self.core.surfaceId, 0)
    end},
    true
)
LuaDefineProperty(
    PlayerPawn.prototype,
    "state",
    {get = function(self)
        return self.core.state
    end},
    true
)
M = {PlayerPawn = PlayerPawn}

return M
